# MNIST CNN implementation using Keras

This project deals with te famous mnist dataset by Yann LeCun. This ipython notebook deals with the implementation of a Convolutional Neural Network(CNN) on mnist data set using Keras deep learning library. The dataset used is taken from Kaggle and is identical to the Yann LeCun's dataset

The first step is to import all the necessary libraries. The use of each library is described as they appear

In [2]:
import numpy as np
import pandas as pd

from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import Flatten
from keras.optimizers import SGD
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D

Using TensorFlow backend.


Pandas is used to load the csv data file

In [4]:
data = pd.read_csv('digits.csv')
data.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8    ...     pixel774  pixel775  pixel776  pixel777  pixel778  \
0       0    ...            0         0         0         0         0   
1       0    ...            0         0         0         0         0   
2       0    ...            0         0         0         0         0   
3       0    ...            0         0         0         0         0   
4       0    ...            0         0         0         0         0   

   pixel779  pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0         0  
1         0         0         0         0         0  
2         0         0         0         0         0  
3         0         0         0         0         0  
4         0         0         0         0         0  

[5 rows x 785 columns]

data contains 785 columns in total. The first column is label and the rest are the pixel values of 28x28 image

The label column is separated and stored separately

In [3]:
label = data['label']

Similarly, all the image pixels are also taken and stored separately

In [4]:
image = data.iloc[0:, 1:]

The label shape is checked. Since its data type is still in the form of pandas datafrome, the second column of shape is not defined and hence we need to reshape it

In [5]:
label.shape

(42000,)

In [6]:
image.shape

(42000, 784)

In [7]:
type(label)

pandas.core.series.Series

In [8]:
type(image)

pandas.core.frame.DataFrame

The above code reveals that the datatype is of pandas dataframe. We need to change its type to numpy array for furthur computations

In [9]:
label = np.asarray(label)

In [10]:
image = np.asarray(image)

Once numpy arrays are obtained, they are reshaped in the required format to recieve the desired output

In [11]:
label = np.reshape(label, [-1, 1])

In [12]:
image = np.reshape(image, [-1, 28, 28, 1])

In [13]:
label.shape

(42000, 1)

In [14]:
image.shape

(42000, 28, 28, 1)

label contains values from 0 to 9. The computations are far more accurate if the labels are converted into their one hot encodings

In [21]:
label = np_utils.to_categorical(label)

In [22]:
label.shape

(42000, 10)

In [26]:
label[0]

array([ 0.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])

Now finally applying the required keras model with necessary dimensional specifications to get the desired output

In [41]:
model = Sequential()
model.add(Conv2D(16, (5, 5), input_shape=(28, 28, 1), padding='same', activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(32, (3, 3), activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(64, (3, 3), padding='same', activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(128, (5, 5), padding='same', activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(1024, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

epochs = 20
lrate = 0.01
sgd = SGD(lr=lrate)
model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

model.fit(image[:40000], label[:40000], epochs=epochs, batch_size=128)

scores = model.evaluate(image[40000:], label[40000:])
print(scores)

Epoch 1/20
40000/40000 [==============================] - 65s - loss: 0.9755 - acc: 0.6745    
Epoch 2/20
40000/40000 [==============================] - 63s - loss: 0.1723 - acc: 0.9510    
Epoch 3/20
40000/40000 [==============================] - 63s - loss: 0.1172 - acc: 0.9654    
Epoch 4/20
40000/40000 [==============================] - 64s - loss: 0.0878 - acc: 0.9745    
Epoch 5/20
40000/40000 [==============================] - 64s - loss: 0.0741 - acc: 0.9792    
Epoch 6/20
40000/40000 [==============================] - 63s - loss: 0.0638 - acc: 0.9822    
Epoch 7/20
40000/40000 [==============================] - 63s - loss: 0.0552 - acc: 0.9843    
Epoch 8/20
40000/40000 [==============================] - 64s - loss: 0.0482 - acc: 0.9860    - ETA: 5s - l
Epoch 9/20
40000/40000 [==============================] - 64s - loss: 0.0420 - acc: 0.9879    
Epoch 10/20
40000/40000 [==============================] - 63s - loss: 0.0367 - acc: 0.9897    
Epoch 11/20
40000/40000 [===========

The accuracy obtained on test data is 98.65%

Now checking the predictions from the model

In [46]:
pred = model.predict(np.reshape(image[0], [1, 28, 28, 1]))

In [47]:
pred

array([[  8.32621645e-12,   1.00000000e+00,   3.28855068e-11,
          2.76061357e-12,   1.44947193e-10,   3.57346201e-11,
          1.55190374e-10,   2.65388156e-10,   4.54934944e-11,
          1.99321601e-10]], dtype=float32)

In [48]:
label[0]

array([ 0.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])

We see the model predicted values correctly

Thank You